### ISSUE TITLE
Support for non-orthogonal slicing (specifically solid transformations)

### ISSUE THREAD

**Issue by jules-vanaret** [napari/napari#7310]
## 🚀 Feature

Currently, a rotated layer via `layer.rotate` is correctly displayed in 3D, but after switching to 2D, the layer is not displayed correctly and following message appears:

`Non-orthogonal slicing is being requested, but is not fully supported. Data is displayed without applying an out-of-slice rotation or shear component.`

I know

- that this has already been asked for in the past (e.g #2616 #3410 #4205)
- that "consistent transformations" is part of the roadmap as per @jni's message in #2616
- that this is a complex problem

I was just wondering

- if a simpler solution could be found specifically for the case of solid transformations (i.e translation + true rotation) 
- if yes, if someone could point me to the relevant part of the code that would need to be modified so that I can try to work on a PR

## Motivation

I am the developer of the plugin [napari-manual-registration](https://github.com/GuignardLab/napari-manual-registration), which allows users to manually find the optimal solid transformation to register two views of the same sample. One of the feature of this plugin is to provide sliders to directly set the values of each transformation component individually. 
When a translation is specified, I can use `layer.translate` which allows for immediate visualization of the result. 
When a rotation is specified, I currently rotate the array itself (!) using `pyclesperanto_prototype.rotate` as a brute force solution to the aforementioned problem. On a good computer with a GPU, the result displays almost instantly, but this is of course not ideal for all other cases. 
This was pointed out by @JoOkuma [here](https://github.com/GuignardLab/napari-manual-registration/issues/2) by the way.

**Comment by jni** [comment]
I think solid and non-solid is not hugely relevant (though I'm not 100% sure) — the main difficulty is orthogonal (which can use standard NumPy slicing e.g. `array[5, :, :]`) and non-orthogonal, which requires materialising and transforming an array of indices and then either using `scipy.ndimage.map_coordinates`, or somehow safely rounding the fancy indexing expression and then using fancy indexing — support for which is not really universal across array types.

Anyway, I was thinking about this and how it's a huge lift, but then I thought, you know what, maybe it's not actually *that* bad and it's a good time to pull off this bandaid.

I recently walked through our image slicing mechanism in [this Zulip comment](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/Will.20napari.20automatically.20resample.20huge.20BMP.20images.20when.20open/near/470065552). I'll copy over the relevant bits for redundancy:

> with *any* array, [napari] first slices the array, then materialises it using `np.asarray`. So, in simplified terms, we do something like `curr_slice = np.asarray(layer.data[5, :, :])` — and this expression works with zarr arrays, dask arrays, numpy arrays...
> 
> […]
> 
> […] the way things work is that (a) we create a SliceRequest [here](https://github.com/napari/napari/blob/15c2d7d5ae7c607e3436800328527bd62c421896/napari/components/_layer_slicer.py#L225), which for Image layers resolves to [this method](https://github.com/napari/napari/blob/15c2d7d5ae7c607e3436800328527bd62c421896/napari/layers/_scalar_field/scalar_field.py#L491). This creates an [_ImageSliceRequest](https://github.com/napari/napari/blob/15c2d7d5ae7c607e3436800328527bd62c421896/napari/layers/image/_slice.py#L165) object which has all the information to slice the arrays. The actual slicing happens in [this line](https://github.com/napari/napari/blob/15c2d7d5ae7c607e3436800328527bd62c421896/napari/layers/image/_slice.py#L217), which calls [this method](https://github.com/napari/napari/blob/15c2d7d5ae7c607e3436800328527bd62c421896/napari/layers/image/_slice.py#L303) on the image layer, where you see the form I pointed out above:
> 
> ```python
>             return np.asarray(data[slices])
> ```
> 
> (underneath you have a bit of extra complexity because of thick slices (#5522)).
> 
> For multi-scales, the code path you follow is [in this part of the _ImageSliceRequest](https://github.com/napari/napari/blob/15c2d7d5ae7c607e3436800328527bd62c421896/napari/layers/image/_slice.py#L234), but it's basically the same — just that it *only* slices the data level needed *and* only for the data that's visible in the canvas currently.

So, having slept on it, how do I see this working? Well, maybe we can:
- grab the slice
- convert it to an array of integer coordinates
- transform the coordinates using the inverse transformation of the image
- (optionally) round the coordinates
- use ndi.map_coordinates or numpy fancy indexing to grab the values
- (optionally) mask values that are out of bounds of the array

We also would need to modify the layer extents, but I think that is also feasible.

There's some undoubtedly tricksy details to be worked out here, such as:

- how much of this works with dask arrays / tensorstore arrays / zarr / xarray arrays?
- do we have to have different branching paths for different array types? Or can we come up with a code path that works for all of them? Or do we just disable this feature when it's not a NumPy array?

Having thought about it for a bit, I'm excited! I am chock-a-block in the next two months but I can probably add a few pair sessions if someone wants to drive this, and/or we can work in December on this stuff.

**Comment by jni** [comment]
btw @jules-vanaret this is one of our oldest feature requests — see #320! 😃 Hence my excitement about having a sketch of a path forward!

**Comment by jules-vanaret** [comment]
Wow! Thank you so much @jni for the detailed answer! :heart_eyes: 
I'll start looking into it. I have a lot on my plate too so this is probably going to be super slow on my end.  

> btw @jules-vanaret this is one of our oldest feature requests — see #320! 😃 Hence my excitement about having a sketch of a path forward!

About the discussion in #320, I see that some users suggested that the issue could be closed by the (at the time) new `.plane` feature of layers. This is more or less the reason I was hoping that implementing non-orthogonal slicing in 2D would be trivial, since the plane view already kinda does that. However, if I am not mistaken, this feature is only compatible with the 3D view, right ? 
Now that I think about it after re-reading #3023, it works in 3D because it implements a custom ray-casting mode, which is not was is used for 2D views as per your comment.

I myself thought a lot about "highjacking" the 2D/3D button in my plugin so that going into 2D mode actually stays in 3D, shows the 3D volumes as planes, and simply rotate the camera so that it faces the plane's normal. This could be a cheapo solution to my problem, although it would require many dirty overrides to make the 3D mode "feel" 2D...

**Comment by jni** [comment]
> However, if I am not mistaken, this feature is only compatible with the 3D view, right ?

> although it would require many dirty overrides to make the 3D mode "feel" 2D...

Yup, all of this. 😂 It could work. But it's also way less efficient – you are loading the entire volume into a texture, rather than the 2D slice. So eventually I think we'll have to do something like my sketch above.

Don't stress about speed, we are a patient bunch. 😂 But do not hesitate if you find a free moment to ask for some help! 😊

**Comment by jni** [comment]
Just dropping a link here that the [ndindex](https://github.com/Quansight-Labs/ndindex) library may be useful here in conjunction with numpy.ix for generating the complete indexing expression before transforming it.

**Comment by sdiebolt** [comment]
Hey! I’d be very interested in helping implement this feature in napari. We’ve found napari’s transform tool to be the best way to initialize volume registrations in the ConfUSIus plugin we're developing. However, because the 2D viewer does not currently support non-orthogonal slicing, it can be difficult to initialize registrations that involve out-of-plane rotations.

I’ve experimented with an agent-assisted implementation and put together a small proof of concept, although I'll need more time to test it more thoroughly and iron out some bugs (and make sure I don't share slop!). At the moment, it only handles the simplest case: solid transformations, no multiscale data or projection mode.

Is there already some ongoing work on this that I could join?

**Comment by jules-vanaret** [comment]
Hey @sdiebolt, thank you for contributing to the discussion! I have been dealing with pretty much the same problem you describe in my own plugin for more than a year now while I postponed working on a fix, so this is definitely much needed!
I wanted to go back to my own pet implementation in a week or so during the Napari hackathon in Krakow to (hopefully) solve this. I don't think I created a specific discussion on Zulip for this when I opened the issue, but we could discuss there if you want to share your progress. :)
